src/scheduler/Scheduler.cpp

```diff
@@ -1129,4 +1197,191 @@ ExecGraphNode Scheduler::getFunctionExecGraphNode(unsigned int messageId)
+            // Second, update our local records
+            pendingMigrations[msgPtr->appid()] = std::move(msgPtr);
+        }
+    }
+}
+
+void Scheduler::broadcastAddPendingMigration(
+  std::shared_ptr<faabric::PendingMigrations> pendingMigrations)
+{
+    // Get all hosts for the to-be migrated app
+    auto msg = pendingMigrations->migrations().at(0).msg();
+    std::string funcStr = faabric::util::funcToString(msg, false);
+    std::set<std::string>& thisRegisteredHosts = registeredHosts[funcStr];
+
+    // Remove this host from the set
+    registeredHosts.erase(thisHost);
+
+    // Send pending migrations to all involved hosts
+    for (auto& otherHost : thisRegisteredHosts) {
+        getFunctionCallClient(otherHost).sendAddPendingMigration(
+          pendingMigrations);
+    }
+}
+
+std::shared_ptr<faabric::PendingMigrations> Scheduler::canAppBeMigrated(
+  uint32_t appId)
+{
+    faabric::util::SharedLock lock(mx);
+
+    if (pendingMigrations.find(appId) == pendingMigrations.end()) {
+        return nullptr;
+    }
+
+    return pendingMigrations[appId];
+}
+
+void Scheduler::addPendingMigration(
+  std::shared_ptr<faabric::PendingMigrations> pMigration)
+{
+    faabric::util::FullLock lock(mx);
+
+    auto msg = pMigration->migrations().at(0).msg();
+    if (pendingMigrations.find(msg.appid()) != pendingMigrations.end()) {
+        SPDLOG_ERROR("Received remote request to add a pending migration for "
+                     "app {}, but already recorded another migration request"
+                     " for the same app.",
+                     msg.appid());
+        throw std::runtime_error("Remote request for app already there");
+    }
+
+    pendingMigrations[msg.appid()] = pMigration;
+}
+
+void Scheduler::removePendingMigration(uint32_t appId)
+{
+    faabric::util::FullLock lock(mx);
+
+    inFlightRequests.erase(appId);
+    pendingMigrations.erase(appId);
+}
+
+std::vector<std::shared_ptr<faabric::PendingMigrations>>
+Scheduler::doCheckForMigrationOpportunities(
+  faabric::util::MigrationStrategy migrationStrategy)
+{
+    std::vector<std::shared_ptr<faabric::PendingMigrations>>
+      pendingMigrationsVec;
+
+    // For each in-flight request that has opted in to be migrated,
+    // check if there is an opportunity to migrate
+    for (const auto& app : inFlightRequests) {
+        auto req = app.second.first;
+        auto originalDecision = *app.second.second;
+
+        // If we have already recorded a pending migration for this req,
+        // skip
+        if (canAppBeMigrated(originalDecision.appId) != nullptr) {
+            SPDLOG_TRACE("Skipping app {} as migration opportunity has "
+                         "already been recorded",
+                         originalDecision.appId);
+            continue;
+        }
+
+        faabric::PendingMigrations msg;
+        msg.set_appid(originalDecision.appId);
+
+        if (migrationStrategy == faabric::util::MigrationStrategy::BIN_PACK) {
```

Comment:
I haven't implemented the `EMPTY_HOSTS` migration strategy, so I haven't implemented support an env. variable to set the migration strategy.